ReactNativeClient/lib/services/KeymapService.ts

```diff
@@ -0,0 +1,317 @@
+				customKeymap.push({ command, accelerator });
+			}
+			return customKeymap;
+		}, []);
+	}
+
+	setKeymap(customKeymap: KeymapItem[]) {
+		this.logger().info('KeymapService: Overriding the keymap...');
+
+		for (let i = 0; i < customKeymap.length; i++) {
+			const item = customKeymap[i];
+
+			try {
+				this.validateKeymapItem(item); // Throws if there are any issues in the keymap item
+				this.setAccelerator(item.command, item.accelerator);
+			} catch (err) {
+				throw new Error(
+					`Keymap item ${JSON.stringify(item)} is invalid!\nReason: ${err.message}`
+				);
+			}
+		}
+
+		try {
+			this.validateKeymap(); // Throws whenever there are duplicate Accelerators used in the keymap
+		} catch (err) {
+			throw new Error(
+				`Keymap configuration contains duplicates!\n${err.message}`
+			);
+		}
+	}
+
+	resetKeymap() {
+		this.keymap = KeymapService.initKeymap();
+	}
+
+	validateKeymapItem(item: KeymapItem) {
+		const commandReference = 'https://github.com/laurent22/joplin/blob/master/README.md';
+		const acceleratorReference = 'https://www.electronjs.org/docs/api/accelerator';
+
+		if (!item.hasOwnProperty('command')) {
+			throw new Error(
+				'Couldn\'t find the "command" property.\n\n' +
+				`Visit ${commandReference} to see the list of available commands.`
+			);
+		} else if (!this.keymap.hasOwnProperty(item.command)) {
+			throw new Error(
+				`"${item.command}" is not a valid command.\n\n` +
+				`Visit ${commandReference} to see the list of available commands.`
+			);
+		}
+
+		if (!item.hasOwnProperty('accelerator')) {
+			throw new Error(
+				'Couldn\'t find the "accelerator" property.\n\n' +
+				`Visit ${acceleratorReference} to learn more about Electron Accelerators.`
+			);
+		} else if (!(item.accelerator === null || KeymapService.isAccelerator(item.accelerator))) {
+			throw new Error(
+				`"${item.accelerator}" is not a valid Accelerator.\n\n` +
+				`Visit ${acceleratorReference} to learn more about Electron Accelerators.`
+			);
+		}
+	}
+
+	validateKeymap() {
+		const contextSets: { [context: string]: Set<string> } = { 'default': new Set() };
+		const contextHasAccelerator = (context: string, accelerator: string) => contextSets[context].has(accelerator);
+
+		for (const item of Object.values(this.keymap)) {
+			const itemAccelerator = item.accelerator;
+			const itemContext = item.context ? item.context : 'default';
+
+			// Accelerators shouldn't be used in "default" context
+			if (itemContext !== 'default' && contextHasAccelerator('default', itemAccelerator)) {
+				const originalItem = Object.values(this.keymap).find(_item => _item.accelerator == item.accelerator);
+
+				this.resetKeymap();
+				throw new Error(
+					`Accelerator "${itemAccelerator}" can't be used for "${item.command}" command because ` +
+					`it's already used in the default context for "${originalItem.command}" command!\n\n` +
+					'You have to change the Accelerator for any of above commands.'
+				);
+			}
```

Comment:
The role of this method, according to its name, is to validate the keymap, so it shouldn't have any side-effects like resetting the keymap. Any side effect should be done where the error is caught (if it is caught somewhere).

Reply:
I agree. I'll add the reset method call to where it's caught.